pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts

```diff
@@ -0,0 +1,162 @@
+/** @jest-environment node */
+
+import fs from 'fs';
+import path from 'path';
+import stream from 'stream';
+
+import { findHomeDir } from '@kubernetes/client-node';
+
+import runCommand, { list } from '@pkg/main/credentialServer/credentialUtils';
+import { spawnFile } from '@pkg/utils/childProcess';
+import paths from '@pkg/utils/paths';
+
+jest.mock('@pkg/utils/childProcess');
+
+describe('runCommand', () => {
+  afterEach(() => {
+    jest.restoreAllMocks();
+    jest.resetAllMocks();
+  });
+
+  it('runs the command', async() => {
+    const expected = `Some output`;
+
+    jest.spyOn(fs.promises, 'readFile').mockImplementation((filepath, options) => {
+      const home = findHomeDir() ?? '';
+
+      expect(filepath).toEqual(path.join(home, '.docker', 'config.json'));
+      expect(options).toMatchObject({ encoding: 'utf-8' });
+
+      return Promise.resolve(JSON.stringify({ credsStore: 'pikachu' }));
+    });
+    jest.mocked(spawnFile).mockImplementation((command, args, options) => {
+      const resourcesPath = path.join(paths.resources, process.platform, 'bin');
+
+      expect(command).toEqual('docker-credential-pikachu');
+      expect(args).toEqual(['pika']);
+      expect(options).toMatchObject({
+        env:   { PATH: expect.stringContaining(resourcesPath) },
+        stdio: [expect.anything(), 'pipe', expect.anything()],
+      });
+
+      return Promise.resolve({ stdout: expected }) as any;
+    });
+    await expect(runCommand('pika')).resolves.toEqual(expected);
+  });
+
+  it('errors out on failing to read config', async() => {
+    const error = new Error('Some error');
+
+    jest.spyOn(fs.promises, 'readFile').mockImplementation((filepath) => {
+      const home = findHomeDir() ?? '';
+
+      expect(filepath).toEqual(path.join(home, '.docker', 'config.json'));
+
+      return Promise.reject(error);
+    });
+
+    jest.mocked(spawnFile).mockImplementation(() => Promise.resolve({}));
+
+    await expect(runCommand('pika')).rejects.toBe(error);
+    expect(jest.mocked(spawnFile)).not.toHaveBeenCalled();
+  });
+
+  describe.each([
```

Comment:
The nested loops obscure what's actually being tested. Could you put some sample tests in the comments above the `describe.each` block so it's clearer.

I wouldn't object to making this a general practice, but in this case the main line of interest is about 25 lines below the start of the first loop.

Reply:
Comments added; hopefully that makes things clearer.